contributors/design-proposals/network/pod-resolv-conf.md

```diff
@@ -0,0 +1,167 @@
+# Custom /etc/resolv.conf
+* Status: pending
+* Version: alpha
+* Implementation owner: Bowei Du <[bowei@google.com](mailto:bowei@google.com)>, Zihong Zheng <[zihongz@google.com](mailto:zihongz@google.com)>
+
+# Overview
+The `/etc/resolv.conf` in a pod is managed by Kubelet and its contents are generated based on `pod.dnsPolicy`. For `dnsPolicy: Default`, `resolv.conf` is copied from the node where the pod is running. If the `dnsPolicy` is `ClusterFirst`, the contents of the resolv.conf is the hosts `resolv.conf` augmented with the following options:
+
+*   Search paths to add aliases for domain names in the same namespace and cluster suffix.
+*   `options ndots` to 5 to ensure the search paths are searched for all potential matches.
+
+The configuration of both search paths and `ndots` results in query amplification of five to ten times for non-cluster internal names. This is due to the fact that each of the search path expansions must be tried before the actual result is found. This order of magnitude increase of query rate imposes a large load on the kube-dns service. At the same time, there are user applications do not need the convenience of the name aliases and do not wish to pay this performance cost.
+
+
+## Existing workarounds
+
+The current work around for this problem is to specify an FQDN for name resolution. Any domain name that ends with a period (e.g. `foo.bar.com.`) will not be search path expanded. However, use of FQDNs is not well-known practice and imposes application-level changes. Cluster operators may not have the luxury of enforcing such a change to applications that run on their infrastructure.
+
+It is also possible for the user to insert a short shell script snippet that rewrites `resolv.conf` on container start-up. This has the same problems as the previous approach and is also awkward for the user. This also forces the container to have additional executable code such as a shell or scripting engine which increases the applications security surface area.
+
+
+# Proposal sketch
+
+Add a new `pod.dnsPolicy: Custom` that allows for user customization of `resolv.conf`.
+
+
+## Pod API example
+
+In the example below, the user wishes to add the pod namespace and a custom expansion to the search path, as they do not use the other name aliases:
+
+```yaml
+# Pod spec
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: ns1
+  name: example
+spec:
+  containers:
+  - name: example
+    image: example
+  dnsPolicy: Custom
+  dnsParams:
```

Comment:
Maybe

```
dnsParams:
  custom:
    search:
```
so we leave room for other params over time?

Reply:
I don't think dnsParams is the best key name for this.  The customization is around resolv.conf, which is used specifically by some resolver implementations.   I think it should be referencing resolv.conf customization.  This should also address the concern that how is this applicable to Windows.

Reply:
@thockin I updated

@phsiao It seems to be asymmetric to have:
```
dnsPolicy: Custom
dnsResolvConf: 
  ...
```
or on Windows
```
dnsPolicy: Custom
dnsWindows: 
  ...
```

The search path expansion seems to be something that all platforms should probably support in some way. As for options, by its nature, I would expect that to be non-portable.

Reply:
How does windows configure name resolution?  We need an answer to that before we merge this, really.

Reply:
Yeah, I want this to be non linux specific.  I really, really, really don't want to embed resolvconf in our API design.  

I'd prefer for people who need to configure resolve.conf themselves supporting a configmap mounted at /etc/resolv.conf.  If we had single file injection we could solve this problem.